File: mini-agent-harness/python/observability_governance.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Awaitable, Literal, Mapping, Protocol


AttemptRoute = Literal["primary", "retry", "fallback"]


@dataclass(frozen=True)
class AttemptIdentity:
    run_id: str
    request_id: str
    attempt_id: str
    route: AttemptRoute


@dataclass(frozen=True)
class TelemetryEvent:
    kind: Literal["attempt_started", "attempt_finished", "tool_finished"]
    event_id: str
    occurred_at_ms: int
    run_id: str
    request_id: str | None = None
    attempt_id: str | None = None
    tool_call_id: str | None = None
    route: AttemptRoute | None = None
    outcome: Literal["succeeded", "failed", "cancelled", "denied"] | None = None
    duration_ms: int | None = None
    ttft_ms: int | None = None
    retry_ordinal: int | None = None


class OpenTelemetryPort(Protocol):
    def export(self, event: TelemetryEvent) -> None | Awaitable[None]: ...


class TelemetryRecorder:
    """Observer failures are counted and never cross into the run outcome."""

    def __init__(self, port: OpenTelemetryPort) -> None:
        self._port = port
        self._observer_failure_count = 0

    async def record(self, event: TelemetryEvent) -> bool:
        try:
            result = self._port.export(event)
            if isinstance(result, Awaitable):
                await result
            return True
        except Exception:
            self._observer_failure_count += 1
            return False

    def report(self) -> Mapping[str, int]:
        return {"observer_failure_count": self._observer_failure_count}


@dataclass(frozen=True)
class UsageSnapshot:
    input_tokens: int
    output_tokens: int
    cache_read_tokens: int = 0
    cache_creation_tokens: int = 0


@dataclass(frozen=True)
class TokenPrice:
    input_usd_per_million: float
    output_usd_per_million: float
    cache_read_usd_per_million: float
    cache_creation_usd_per_million: float


@dataclass(frozen=True)
class UsageCostCommand:
    event_id: str
    identity: AttemptIdentity
    model: str
    price_version: str
    captured_at_ms: int
    cumulative: UsageSnapshot
    price: TokenPrice | None = None
    ttft_ms: int | None = None


@dataclass(frozen=True)
class UsageCostEntry:
    event_id: str
    identity: AttemptIdentity
    model: str
    price_version: str
    captured_at_ms: int
    cumulative: UsageSnapshot
    delta: UsageSnapshot
    cost_status: Literal["known", "unknown"]
    cost_usd: float | None
    ttft_status: Literal["observed", "unknown"]
    ttft_ms: int | None


class EventCollisionError(ValueError):
    pass


class NonMonotonicUsageError(ValueError):
    pass


class UsageCostLedger:
    def __init__(self) -> None:
        self._entries: list[UsageCostEntry] = []
        self._events: dict[str, tuple[UsageCostCommand, UsageCostEntry]] = {}
        self._last_by_attempt: dict[tuple[str, str, str], UsageSnapshot] = {}

    def record(self, command: UsageCostCommand) -> UsageCostEntry:
        _validate_usage_command(command)
        existing = self._events.get(command.event_id)
        if existing is not None:
            if existing[0] != command:
                raise EventCollisionError(f"usage event id collision: {command.event_id}")
            return existing[1]

        key = (
            command.identity.run_id,
            command.identity.request_id,
            command.identity.attempt_id,
        )
        previous = self._last_by_attempt.get(key, UsageSnapshot(0, 0, 0, 0))
        delta = _subtract_usage(command.cumulative, previous)
        cost = _calculate_cost(delta, command.price) if command.price else None
        entry = UsageCostEntry(
            command.event_id,
            command.identity,
            command.model,
            command.price_version,
            command.captured_at_ms,
            command.cumulative,
            delta,
            "known" if command.price else "unknown",
            cost,
            "observed" if command.ttft_ms is not None else "unknown",
            command.ttft_ms,
        )
        self._last_by_attempt[key] = command.cumulative
        self._entries.append(entry)
        self._events[command.event_id] = (command, entry)
        return entry

    def entries(self) -> tuple[UsageCostEntry, ...]:
        return tuple(self._entries)

    def report(self) -> Mapping[str, int | float]:
        return {
            "entry_count": len(self._entries),
            "known_cost_usd": sum(entry.cost_usd or 0 for entry in self._entries),
            "unknown_cost_entry_count": sum(
                entry.cost_status == "unknown" for entry in self._entries
            ),
            "unknown_ttft_entry_count": sum(
                entry.ttft_status == "unknown" for entry in self._entries
            ),
        }


EvaluationName = Literal["correctness", "grounding", "safety", "efficiency"]


@dataclass(frozen=True)
class EvaluationDimension:
    name: EvaluationName
    score: float
    passed: bool


@dataclass(frozen=True)
class EvaluationRecord:
    event_id: str
    run_id: str
    evaluation_id: str
    rubric_id: str
    rubric_version: int
    evaluator_version: str
    outcome: Literal["passed", "failed", "inconclusive"]
    dimensions: tuple[EvaluationDimension, ...]
    recorded_at_ms: int


class EvaluationLedger:
    def __init__(self) -> None:
        self._records: list[EvaluationRecord] = []
        self._events: dict[str, EvaluationRecord] = {}

    def record(self, record: EvaluationRecord) -> EvaluationRecord:
        _require_text(record.event_id, "event_id")
        _require_text(record.rubric_id, "rubric_id")
        if record.rubric_version < 1:
            raise ValueError("rubric_version must be positive")
        existing = self._events.get(record.event_id)
        if existing is not None:
            if existing != record:
                raise EventCollisionError(
                    f"evaluation event id collision: {record.event_id}"
                )
            return existing
        self._events[record.event_id] = record
        self._records.append(record)
        return record

    def records(self) -> tuple[EvaluationRecord, ...]:
        return tuple(self._records)


@dataclass(frozen=True)
class TenantLimit:
    max_tokens_per_window: int
    max_cost_usd_per_window: float
    max_concurrency: int
    max_queue_size: int


@dataclass(frozen=True)
class ReservationRequest:
    tenant_id: str
    reservation_id: str
    estimated_tokens: int
    estimated_cost_usd: float
    requested_at_ms: int


@dataclass(frozen=True)
class TenantReservation:
    tenant_id: str
    reservation_id: str
    estimated_tokens: int
    estimated_cost_usd: float
    requested_at_ms: int
    admitted_at_ms: int


@dataclass(frozen=True)
class TenantGovernanceReport:
    tenant_id: str
    active_concurrency: int
    queue_depth: int
    reserved_tokens: int
    reserved_cost_usd: float
    consumed_tokens: int
    consumed_cost_usd: float


class UnknownTenantError(KeyError):
    pass


class TenantQueueFullError(RuntimeError):
    pass


class ReservationTooLargeError(ValueError):
    pass


class ReservationStateError(RuntimeError):
    pass


class ReservationCancelledError(asyncio.CancelledError):
    pass


@dataclass
class _QueuedReservation:
    request: ReservationRequest
    future: asyncio.Future[TenantReservation]


@dataclass
class _TenantState:
    limit: TenantLimit
    active: dict[str, TenantReservation]
    queue: list[_QueuedReservation]
    known_ids: set[str]
    consumed_tokens: int = 0
    consumed_cost_usd: float = 0.0


class TenantGovernor:
    """In-process reference governor; distributed CAS and fairness are deferred."""

    def __init__(self, limits: Mapping[str, TenantLimit], now=lambda: 0) -> None:
        self._now = now
        self._states: dict[str, _TenantState] = {}
        for tenant_id, limit in limits.items():
            _validate_limit(limit)
            self._states[tenant_id] = _TenantState(limit, {}, [], set())

    async def reserve(self, request: ReservationRequest) -> TenantReservation:
        _validate_reservation(request)
        state = self._state(request.tenant_id)
        if request.reservation_id in state.known_ids:
            raise ReservationStateError(
                f"duplicate reservation id: {request.reservation_id}"
            )
        if (
            request.estimated_tokens > state.limit.max_tokens_per_window
            or request.estimated_cost_usd > state.limit.max_cost_usd_per_window
        ):
            raise ReservationTooLargeError(
                f"reservation exceeds tenant limit: {request.reservation_id}"
            )
        state.known_ids.add(request.reservation_id)
        if _can_admit(state, request):
            return self._admit(state, request)
        if len(state.queue) >= state.limit.max_queue_size:
            state.known_ids.remove(request.reservation_id)
            raise TenantQueueFullError(f"tenant queue full: {request.tenant_id}")
        future = asyncio.get_running_loop().create_future()
        state.queue.append(_QueuedReservation(request, future))
        return await future

    def complete(
        self,
        tenant_id: str,
        reservation_id: str,
        actual_tokens: int,
        actual_cost_usd: float,
    ) -> TenantGovernanceReport:
        _validate_amount(actual_tokens, "actual_tokens", integer=True)
        _validate_amount(actual_cost_usd, "actual_cost_usd")
        state = self._state(tenant_id)
        if state.active.pop(reservation_id, None) is None:
            raise ReservationStateError(
                f"reservation is not active: {reservation_id}"
            )
        state.consumed_tokens += actual_tokens
        state.consumed_cost_usd += actual_cost_usd
        self._promote(state)
        return _report(tenant_id, state)

    def cancel(self, tenant_id: str, reservation_id: str) -> TenantGovernanceReport:
        state = self._state(tenant_id)
        if state.active.pop(reservation_id, None) is None:
            index = next(
                (
                    index
                    for index, item in enumerate(state.queue)
                    if item.request.reservation_id == reservation_id
                ),
                -1,
            )
            if index < 0:
                raise ReservationStateError(
                    f"reservation is not active or queued: {reservation_id}"
                )
            queued = state.queue.pop(index)
            queued.future.set_exception(
                ReservationCancelledError(f"reservation cancelled: {reservation_id}")
            )
        self._promote(state)
        return _report(tenant_id, state)

    def reset_budget_window(self, tenant_id: str) -> TenantGovernanceReport:
        state = self._state(tenant_id)
        state.consumed_tokens = 0
        state.consumed_cost_usd = 0
        self._promote(state)
        return _report(tenant_id, state)

    def report(self, tenant_id: str) -> TenantGovernanceReport:
        return _report(tenant_id, self._state(tenant_id))

    def _state(self, tenant_id: str) -> _TenantState:
        try:
            return self._states[tenant_id]
        except KeyError as error:
            raise UnknownTenantError(f"unknown tenant: {tenant_id}") from error

    def _admit(
        self, state: _TenantState, request: ReservationRequest
    ) -> TenantReservation:
        reservation = TenantReservation(
            request.tenant_id,
            request.reservation_id,
            request.estimated_tokens,
            request.estimated_cost_usd,
            request.requested_at_ms,
            self._now(),
        )
        state.active[request.reservation_id] = reservation
        return reservation

    def _promote(self, state: _TenantState) -> None:
        while state.queue and _can_admit(state, state.queue[0].request):
            queued = state.queue.pop(0)
            queued.future.set_result(self._admit(state, queued.request))


def _subtract_usage(current: UsageSnapshot, previous: UsageSnapshot) -> UsageSnapshot:
    values = (
        current.input_tokens - previous.input_tokens,
        current.output_tokens - previous.output_tokens,
        current.cache_read_tokens - previous.cache_read_tokens,
        current.cache_creation_tokens - previous.cache_creation_tokens,
    )
    if any(value < 0 for value in values):
        raise NonMonotonicUsageError("cumulative usage decreased")
    return UsageSnapshot(*values)


def _calculate_cost(usage: UsageSnapshot, price: TokenPrice) -> float:
    return (
        usage.input_tokens * price.input_usd_per_million
        + usage.output_tokens * price.output_usd_per_million
        + usage.cache_read_tokens * price.cache_read_usd_per_million
        + usage.cache_creation_tokens * price.cache_creation_usd_per_million
    ) / 1_000_000


def _validate_usage_command(command: UsageCostCommand) -> None:
    _require_text(command.event_id, "event_id")
    _require_text(command.identity.run_id, "run_id")
    _require_text(command.identity.request_id, "request_id")
    _require_text(command.identity.attempt_id, "attempt_id")
    _require_text(command.model, "model")
    _require_text(command.price_version, "price_version")
    for name, value in vars(command.cumulative).items():
        _validate_amount(value, name, integer=True)
    if command.ttft_ms is not None:
        _validate_amount(command.ttft_ms, "ttft_ms", integer=True)


def _validate_limit(limit: TenantLimit) -> None:
    _validate_amount(limit.max_tokens_per_window, "max_tokens_per_window", integer=True)
    _validate_amount(limit.max_cost_usd_per_window, "max_cost_usd_per_window")
    if limit.max_concurrency < 1:
        raise ValueError("max_concurrency must be positive")
    if limit.max_queue_size < 0:
        raise ValueError("max_queue_size must be non-negative")


def _validate_reservation(request: ReservationRequest) -> None:
    _require_text(request.tenant_id, "tenant_id")
    _require_text(request.reservation_id, "reservation_id")
    _validate_amount(request.estimated_tokens, "estimated_tokens", integer=True)
    _validate_amount(request.estimated_cost_usd, "estimated_cost_usd")


def _validate_amount(value: int | float, name: str, integer: bool = False) -> None:
    if value < 0 or (integer and not isinstance(value, int)):
        raise ValueError(f"{name} must be a non-negative {'integer' if integer else 'number'}")


def _require_text(value: str, name: str) -> None:
    if not value.strip():
        raise ValueError(f"{name} must not be empty")


def _can_admit(state: _TenantState, request: ReservationRequest) -> bool:
    reserved_tokens = sum(item.estimated_tokens for item in state.active.values())
    reserved_cost = sum(item.estimated_cost_usd for item in state.active.values())
    return (
        len(state.active) < state.limit.max_concurrency
        and state.consumed_tokens + reserved_tokens + request.estimated_tokens
        <= state.limit.max_tokens_per_window
        and state.consumed_cost_usd + reserved_cost + request.estimated_cost_usd
        <= state.limit.max_cost_usd_per_window
    )


def _report(tenant_id: str, state: _TenantState) -> TenantGovernanceReport:
    return TenantGovernanceReport(
        tenant_id,
        len(state.active),
        len(state.queue),
        sum(item.estimated_tokens for item in state.active.values()),
        sum(item.estimated_cost_usd for item in state.active.values()),
        state.consumed_tokens,
        state.consumed_cost_usd,
    )

File: mini-agent-harness/typescript/agent/observabilityGovernance.ts
export type AttemptRoute = 'primary' | 'retry' | 'fallback'

export type AttemptIdentity = Readonly<{
  runId: string
  requestId: string
  attemptId: string
  route: AttemptRoute
}>

export type TelemetryEvent =
  | Readonly<{
      kind: 'attempt_started'
      eventId: string
      identity: AttemptIdentity
      occurredAtMs: number
      retryOrdinal: number
    }>
  | Readonly<{
      kind: 'attempt_finished'
      eventId: string
      identity: AttemptIdentity
      occurredAtMs: number
      outcome: 'succeeded' | 'failed' | 'cancelled'
      durationMs: number
      ttftMs?: number
    }>
  | Readonly<{
      kind: 'tool_finished'
      eventId: string
      runId: string
      toolCallId: string
      occurredAtMs: number
      outcome: 'succeeded' | 'failed' | 'cancelled' | 'denied'
      durationMs: number
    }>

export interface OpenTelemetryPort {
  export(event: TelemetryEvent): void | Promise<void>
}

/** Observer failures are counted and never cross into the run outcome. */
export class TelemetryRecorder {
  #observerFailureCount = 0
  readonly #port: OpenTelemetryPort

  constructor(port: OpenTelemetryPort) {
    this.#port = port
  }

  async record(event: TelemetryEvent): Promise<boolean> {
    try {
      await this.#port.export(deepFreeze(structuredClone(event)))
      return true
    } catch {
      this.#observerFailureCount++
      return false
    }
  }

  report(): Readonly<{ observerFailureCount: number }> {
    return Object.freeze({ observerFailureCount: this.#observerFailureCount })
  }
}

export type UsageSnapshot = Readonly<{
  inputTokens: number
  outputTokens: number
  cacheReadTokens: number
  cacheCreationTokens: number
}>

export type TokenPrice = Readonly<{
  inputUsdPerMillion: number
  outputUsdPerMillion: number
  cacheReadUsdPerMillion: number
  cacheCreationUsdPerMillion: number
}>

export type UsageCostCommand = Readonly<{
  eventId: string
  identity: AttemptIdentity
  model: string
  priceVersion: string
  capturedAtMs: number
  cumulative: UsageSnapshot
  price?: TokenPrice
  ttftMs?: number
}>

export type CostState =
  | Readonly<{ status: 'known'; usd: number }>
  | Readonly<{ status: 'unknown' }>

export type TtftState =
  | Readonly<{ status: 'observed'; milliseconds: number }>
  | Readonly<{ status: 'unknown' }>

export type UsageCostEntry = Readonly<{
  eventId: string
  identity: AttemptIdentity
  model: string
  priceVersion: string
  capturedAtMs: number
  cumulative: UsageSnapshot
  delta: UsageSnapshot
  cost: CostState
  ttft: TtftState
}>

export class EventCollisionError extends Error {}
export class NonMonotonicUsageError extends Error {}

/** Converts provider cumulative snapshots into attempt-local immutable deltas. */
export class UsageCostLedger {
  readonly #entries: UsageCostEntry[] = []
  readonly #events = new Map<string, Readonly<{ fingerprint: string; entry: UsageCostEntry }>>()
  readonly #lastByAttempt = new Map<string, UsageSnapshot>()

  record(command: UsageCostCommand): UsageCostEntry {
    validateUsageCommand(command)
    const fingerprint = usageFingerprint(command)
    const existing = this.#events.get(command.eventId)
    if (existing) {
      if (existing.fingerprint !== fingerprint) {
        throw new EventCollisionError(`usage event id collision: ${command.eventId}`)
      }
      return existing.entry
    }

    const attemptKey = identityKey(command.identity)
    const previous = this.#lastByAttempt.get(attemptKey) ?? zeroUsage()
    const delta = subtractUsage(command.cumulative, previous)
    const entry: UsageCostEntry = deepFreeze({
      eventId: command.eventId,
      identity: structuredClone(command.identity),
      model: command.model,
      priceVersion: command.priceVersion,
      capturedAtMs: command.capturedAtMs,
      cumulative: structuredClone(command.cumulative),
      delta,
      cost: command.price
        ? { status: 'known', usd: calculateCost(delta, command.price) }
        : { status: 'unknown' },
      ttft: command.ttftMs === undefined
        ? { status: 'unknown' }
        : { status: 'observed', milliseconds: command.ttftMs },
    })
    this.#lastByAttempt.set(attemptKey, entry.cumulative)
    this.#entries.push(entry)
    this.#events.set(command.eventId, Object.freeze({ fingerprint, entry }))
    return entry
  }

  entries(): readonly UsageCostEntry[] {
    return Object.freeze([...this.#entries])
  }

  report(): Readonly<{
    entryCount: number
    knownCostUsd: number
    unknownCostEntryCount: number
    unknownTtftEntryCount: number
  }> {
    return Object.freeze({
      entryCount: this.#entries.length,
      knownCostUsd: this.#entries.reduce(
        (total, entry) => total + (entry.cost.status === 'known' ? entry.cost.usd : 0),
        0,
      ),
      unknownCostEntryCount: this.#entries.filter(entry => entry.cost.status === 'unknown').length,
      unknownTtftEntryCount: this.#entries.filter(entry => entry.ttft.status === 'unknown').length,
    })
  }
}

export type EvaluationDimension = Readonly<{
  name: 'correctness' | 'grounding' | 'safety' | 'efficiency'
  score: number
  passed: boolean
}>

export type EvaluationRecord = Readonly<{
  eventId: string
  runId: string
  evaluationId: string
  rubricId: string
  rubricVersion: number
  evaluatorVersion: string
  outcome: 'passed' | 'failed' | 'inconclusive'
  dimensions: readonly EvaluationDimension[]
  recordedAtMs: number
}>

export class EvaluationLedger {
  readonly #records: EvaluationRecord[] = []
  readonly #fingerprints = new Map<string, string>()

  record(input: EvaluationRecord): EvaluationRecord {
    validateEvaluation(input)
    const frozen = deepFreeze(structuredClone(input))
    const fingerprint = evaluationFingerprint(frozen)
    const existing = this.#fingerprints.get(input.eventId)
    if (existing !== undefined) {
      if (existing !== fingerprint) {
        throw new EventCollisionError(`evaluation event id collision: ${input.eventId}`)
      }
      return this.#records.find(record => record.eventId === input.eventId)!
    }
    this.#fingerprints.set(input.eventId, fingerprint)
    this.#records.push(frozen)
    return frozen
  }

  records(): readonly EvaluationRecord[] {
    return Object.freeze([...this.#records])
  }
}

export type TenantLimit = Readonly<{
  maxTokensPerWindow: number
  maxCostUsdPerWindow: number
  maxConcurrency: number
  maxQueueSize: number
}>

export type ReservationRequest = Readonly<{
  tenantId: string
  reservationId: string
  estimatedTokens: number
  estimatedCostUsd: number
  requestedAtMs: number
}>

export type TenantReservation = ReservationRequest & Readonly<{
  admittedAtMs: number
}>

export type TenantGovernanceReport = Readonly<{
  tenantId: string
  activeConcurrency: number
  queueDepth: number
  reservedTokens: number
  reservedCostUsd: number
  consumedTokens: number
  consumedCostUsd: number
}>

export class UnknownTenantError extends Error {}
export class TenantQueueFullError extends Error {}
export class ReservationTooLargeError extends Error {}
export class ReservationStateError extends Error {}
export class ReservationCancelledError extends Error {}

type QueuedReservation = Readonly<{
  request: ReservationRequest
  resolve: (reservation: TenantReservation) => void
  reject: (error: Error) => void
}>

type TenantState = {
  readonly limit: TenantLimit
  readonly active: Map<string, TenantReservation>
  readonly queue: QueuedReservation[]
  readonly knownIds: Set<string>
  consumedTokens: number
  consumedCostUsd: number
}

/** In-process reference governor. Distributed CAS and cross-node fairness are deferred. */
export class TenantGovernor {
  readonly #states = new Map<string, TenantState>()
  readonly #now: () => number

  constructor(limits: Readonly<Record<string, TenantLimit>>, now: () => number = Date.now) {
    this.#now = now
    for (const [tenantId, limit] of Object.entries(limits)) {
      validateTenantLimit(limit)
      this.#states.set(tenantId, {
        limit: deepFreeze(structuredClone(limit)),
        active: new Map(),
        queue: [],
        knownIds: new Set(),
        consumedTokens: 0,
        consumedCostUsd: 0,
      })
    }
  }

  reserve(input: ReservationRequest): Promise<TenantReservation> {
    validateReservation(input)
    const state = this.#state(input.tenantId)
    if (state.knownIds.has(input.reservationId)) {
      throw new ReservationStateError(`duplicate reservation id: ${input.reservationId}`)
    }
    if (
      input.estimatedTokens > state.limit.maxTokensPerWindow
      || input.estimatedCostUsd > state.limit.maxCostUsdPerWindow
    ) {
      throw new ReservationTooLargeError(`reservation exceeds tenant limit: ${input.reservationId}`)
    }
    state.knownIds.add(input.reservationId)
    if (canAdmit(state, input)) {
      return Promise.resolve(this.#admit(state, input))
    }
    if (state.queue.length >= state.limit.maxQueueSize) {
      state.knownIds.delete(input.reservationId)
      throw new TenantQueueFullError(`tenant queue full: ${input.tenantId}`)
    }
    return new Promise<TenantReservation>((resolve, reject) => {
      state.queue.push(Object.freeze({ request: deepFreeze(structuredClone(input)), resolve, reject }))
    })
  }

  complete(
    tenantId: string,
    reservationId: string,
    actualTokens: number,
    actualCostUsd: number,
  ): TenantGovernanceReport {
    validateAmount(actualTokens, 'actualTokens', true)
    validateAmount(actualCostUsd, 'actualCostUsd')
    const state = this.#state(tenantId)
    if (!state.active.delete(reservationId)) {
      throw new ReservationStateError(`reservation is not active: ${reservationId}`)
    }
    state.consumedTokens += actualTokens
    state.consumedCostUsd += actualCostUsd
    this.#promote(state)
    return reportForTenant(tenantId, state)
  }

  cancel(tenantId: string, reservationId: string): TenantGovernanceReport {
    const state = this.#state(tenantId)
    if (!state.active.delete(reservationId)) {
      const index = state.queue.findIndex(item => item.request.reservationId === reservationId)
      if (index < 0) throw new ReservationStateError(`reservation is not active or queued: ${reservationId}`)
      const [queued] = state.queue.splice(index, 1)
      queued!.reject(new ReservationCancelledError(`reservation cancelled: ${reservationId}`))
    }
    this.#promote(state)
    return reportForTenant(tenantId, state)
  }

  resetBudgetWindow(tenantId: string): TenantGovernanceReport {
    const state = this.#state(tenantId)
    state.consumedTokens = 0
    state.consumedCostUsd = 0
    this.#promote(state)
    return reportForTenant(tenantId, state)
  }

  report(tenantId: string): TenantGovernanceReport {
    return reportForTenant(tenantId, this.#state(tenantId))
  }

  #state(tenantId: string): TenantState {
    const state = this.#states.get(tenantId)
    if (!state) throw new UnknownTenantError(`unknown tenant: ${tenantId}`)
    return state
  }

  #admit(state: TenantState, input: ReservationRequest): TenantReservation {
    const reservation = deepFreeze({ ...structuredClone(input), admittedAtMs: this.#now() })
    state.active.set(input.reservationId, reservation)
    return reservation
  }

  #promote(state: TenantState): void {
    while (state.queue.length > 0 && canAdmit(state, state.queue[0]!.request)) {
      const queued = state.queue.shift()!
      queued.resolve(this.#admit(state, queued.request))
    }
  }
}

function identityKey(identity: AttemptIdentity): string {
  return `${identity.runId}\u0000${identity.requestId}\u0000${identity.attemptId}`
}

function zeroUsage(): UsageSnapshot {
  return Object.freeze({ inputTokens: 0, outputTokens: 0, cacheReadTokens: 0, cacheCreationTokens: 0 })
}

function subtractUsage(current: UsageSnapshot, previous: UsageSnapshot): UsageSnapshot {
  const fields = ['inputTokens', 'outputTokens', 'cacheReadTokens', 'cacheCreationTokens'] as const
  const delta = Object.fromEntries(fields.map(field => {
    if (current[field] < previous[field]) {
      throw new NonMonotonicUsageError(`cumulative usage decreased for ${field}`)
    }
    return [field, current[field] - previous[field]]
  })) as UsageSnapshot
  return Object.freeze(delta)
}

function calculateCost(usage: UsageSnapshot, price: TokenPrice): number {
  return (
    usage.inputTokens * price.inputUsdPerMillion
    + usage.outputTokens * price.outputUsdPerMillion
    + usage.cacheReadTokens * price.cacheReadUsdPerMillion
    + usage.cacheCreationTokens * price.cacheCreationUsdPerMillion
  ) / 1_000_000
}

function usageFingerprint(command: UsageCostCommand): string {
  return JSON.stringify({
    eventId: command.eventId,
    identity: command.identity,
    model: command.model,
    priceVersion: command.priceVersion,
    capturedAtMs: command.capturedAtMs,
    cumulative: command.cumulative,
    price: command.price ?? null,
    ttftMs: command.ttftMs ?? null,
  })
}

function evaluationFingerprint(record: EvaluationRecord): string {
  return JSON.stringify(record)
}

function validateUsageCommand(command: UsageCostCommand): void {
  requireText(command.eventId, 'eventId')
  requireText(command.identity.runId, 'runId')
  requireText(command.identity.requestId, 'requestId')
  requireText(command.identity.attemptId, 'attemptId')
  requireText(command.model, 'model')
  requireText(command.priceVersion, 'priceVersion')
  for (const [name, value] of Object.entries(command.cumulative)) validateAmount(value, name, true)
  if (command.ttftMs !== undefined) validateAmount(command.ttftMs, 'ttftMs')
  if (command.price) {
    for (const [name, value] of Object.entries(command.price)) validateAmount(value, name)
  }
}

function validateEvaluation(record: EvaluationRecord): void {
  requireText(record.eventId, 'eventId')
  requireText(record.runId, 'runId')
  requireText(record.evaluationId, 'evaluationId')
  requireText(record.rubricId, 'rubricId')
  requireText(record.evaluatorVersion, 'evaluatorVersion')
  if (!Number.isInteger(record.rubricVersion) || record.rubricVersion < 1) {
    throw new Error('rubricVersion must be a positive integer')
  }
  for (const dimension of record.dimensions) {
    if (!Number.isFinite(dimension.score)) throw new Error('evaluation score must be finite')
  }
}

function validateTenantLimit(limit: TenantLimit): void {
  validateAmount(limit.maxTokensPerWindow, 'maxTokensPerWindow', true)
  validateAmount(limit.maxCostUsdPerWindow, 'maxCostUsdPerWindow')
  if (!Number.isInteger(limit.maxConcurrency) || limit.maxConcurrency < 1) {
    throw new Error('maxConcurrency must be a positive integer')
  }
  if (!Number.isInteger(limit.maxQueueSize) || limit.maxQueueSize < 0) {
    throw new Error('maxQueueSize must be a non-negative integer')
  }
}

function validateReservation(input: ReservationRequest): void {
  requireText(input.tenantId, 'tenantId')
  requireText(input.reservationId, 'reservationId')
  validateAmount(input.estimatedTokens, 'estimatedTokens', true)
  validateAmount(input.estimatedCostUsd, 'estimatedCostUsd')
}

function validateAmount(value: number, name: string, integer = false): void {
  if (!Number.isFinite(value) || value < 0 || (integer && !Number.isInteger(value))) {
    throw new Error(`${name} must be a non-negative ${integer ? 'integer' : 'number'}`)
  }
}

function requireText(value: string, name: string): void {
  if (value.trim() === '') throw new Error(`${name} must not be empty`)
}

function canAdmit(state: TenantState, input: ReservationRequest): boolean {
  const reservedTokens = [...state.active.values()].reduce((sum, item) => sum + item.estimatedTokens, 0)
  const reservedCost = [...state.active.values()].reduce((sum, item) => sum + item.estimatedCostUsd, 0)
  return state.active.size < state.limit.maxConcurrency
    && state.consumedTokens + reservedTokens + input.estimatedTokens <= state.limit.maxTokensPerWindow
    && state.consumedCostUsd + reservedCost + input.estimatedCostUsd <= state.limit.maxCostUsdPerWindow
}

function reportForTenant(tenantId: string, state: TenantState): TenantGovernanceReport {
  return Object.freeze({
    tenantId,
    activeConcurrency: state.active.size,
    queueDepth: state.queue.length,
    reservedTokens: [...state.active.values()].reduce((sum, item) => sum + item.estimatedTokens, 0),
    reservedCostUsd: [...state.active.values()].reduce((sum, item) => sum + item.estimatedCostUsd, 0),
    consumedTokens: state.consumedTokens,
    consumedCostUsd: state.consumedCostUsd,
  })
}

function deepFreeze<T>(value: T): T {
  if (value && typeof value === 'object' && !Object.isFrozen(value)) {
    Object.freeze(value)
    for (const child of Object.values(value)) deepFreeze(child)
  }
  return value
}
